### ISSUE TITLE
Scheduler gap in stream_uv?

### ISSUE THREAD

**Issue by tro3** [JuliaLang/julia#38992]
Ran across this while working on #38916.  Summary: when scheduling Tasks and Channels inside of a testset, it is possible to break the low-level Julia task scheduling scheme somehow.  I am going to try to fix this via PR, but crawling through the low-level innards without familiarity is a slog - someone who knows the code better than I will likely see the problem before I do, so I figured I'd submit this to help shorten the time.

MWE is attached below.  When run, it generates the following error (line numbers here for Julia 1.5.3/win64, but same error occurs on master/win64 and on all platforms on the CI buildbot):

```julia
julia> include("script.jl")
Starting Task 1
Starting Task 2
Test Summary: |ERROR: LoadError: TypeError: in typeassert, expected Int32, got a value of type Nothing
Stacktrace:
 [1] uv_write(::Base.TTY, ::Ptr{UInt8}, ::UInt64) at .\stream.jl:933
 [2] unsafe_write(::Base.TTY, ::Ptr{UInt8}, ::UInt64) at .\stream.jl:1005
...
```

If Julia's scheduler is holding up, failing this type assertion shouldn't be possible.  The low-level async call stack behind this statement (started from a simple `print` statement in `print_test_results` in Test.jl) looks like:

* `uv_write` (stream.jl)
    * starts with iolock
    * calls `uv_write_async` (stream.jl)
        * allocates `uvw` (a uv request)
        * calls `:jl_uv_write` (jl_uv.c) with `uvw` as request and `uv_writecb_task` as the callback
        * returns `uvw`
    * assigns current task id to `uvw` data
    * releases iolock
    * `wait()`s to be rescheduled, with the asserted `wait()` return value as the `uv_write` result status
*  `uv_writecb_task`  (stream.jl), running when `jl_uv_write` above finishes
    * Gets the orginal task from the `uvw` data
    * Schedules the orginal task with the `uv_write` result status as the `wait()` return value

So with every `print` statement, a libuv + callback Task is launched, with the callback re-`schedule`ing the original Task and passing the return value.  But somehow in the MWE, the original Task is getting re-scheduled from _somewhere else_ with no return value, and this is triggering the exception.  Tasks may only be scheduled by one other item - if something else had scheduled the original Task first, the callback would see an exception and terminate.  But I've tested for this condition and not seen it.

A few more important points:
* I have not yet been able to recreate outside of the Test environment.  It always seems to trigger on a Test result write.
* The throwing of the exception in the subtask is required.  If the tasks end cleanly, the issue goes away.
* If the Channel is moved _inside_ the function boundary, the issue goes away.  I have not yet dug into the Channel mechanics, but that is where I'll be burrowing next.  The fact that the Channel has to be a closure or global (not sure which it is inside a Test block), might move this issue to the compiler level, in which case I'll be stuck.

Anyway, I'll keep working on this to try to find the weakness, but any help or advice is appreciated.  

MWE:

```julia
using Test

@testset "MWE" begin

    c = Channel(0)

    function f()
        #c = Channel(0)        # Passes if here
        ct = current_task()
        t1 = @async try
                println("Starting Task 1")
                put!(c,1)
            catch
                rethrow()
            finally
                ct.queue === nothing && schedule(ct)
            end
        t2 = @async try
                println("Starting Task 2")
                undefined()
                r = take!(c)
            catch
                rethrow()
            finally
                ct.queue === nothing && schedule(ct)
            end
        ts = [t1,t2]
        while true
            wait()
            if any(istaskfailed, ts)
                error("Task failed")
            elseif all(istaskdone, ts)
                break
            end
        end
        #close(c)        # Passes if here
    end

    ex = nothing
    try
        f()
    catch e
        ex = e
    end

    close(c)

end
```

**Comment by vtjnash** [comment]
```
            finally
                ct.queue === nothing && schedule(ct)
```
This seems quite certain to occasionally corrupt the scheduler, since it's conditionally scheduling a Task to receive `nothing` that it's expecting never to get. The rate depends on when `yield` gets called elsewhere, or how many threads there are.

**Comment by tro3** [comment]
1. I think you are right in that _that_ is the source of the mystery schedule.  As is, it is being executed in the main Task, which is likely where the `print` is coming from.  If I replace `f()` with `fetch(schedule(Task(f)))`, the issue goes away. 
2. But if so, it implies one can never `schedule` anything when there are `print` statements running around.  That `uv_write` implementation fundamentally assumes nothing else will ever schedule the calling Task.  Seems like a dangerous assumption.

A cleaner implementation would to have the calling Task wait on a callback notification specifically, then get the return value through a Channel (or something more optimized as that seems overkill).  @vtjnash Is this worth working on?

**Comment by tro3** [comment]
Oh, I see.  The `schedule(taker, result)` is used in Channels as well.  So the above scheme won't work.  So basically, one should only `schedule` Tasks under their control.  I'll chew on this some more, but I can live with the restriction.  Thanks.

**Comment by vtjnash** [comment]
Yes, `schedule` is the primitive, so there's limited capability for it to do otherwise. But accordingly, most users (who aren't experimenting with alternative `@sync` implementations lol) should hopefully be using higher level constructs instead. But maybe this is close to the behavior you were wanting?
```julia
julia> Base.Experimental.@sync begin
         @async wait()
         @async error("die")
         @async wait()
       end
ERROR: TaskFailedException
Stacktrace:
 [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
   @ Base ./task.jl:705
 [2] wait
   @ ./task.jl:764 [inlined]
 [3] wait(c::Base.GenericCondition{ReentrantLock})
   @ Base ./condition.jl:106
 [4] take_buffered(c::Channel{Any})
   @ Base ./channels.jl:389
 [5] take!
   @ ./channels.jl:383 [inlined]
 [6] sync_end(c::Channel{Any})
   @ Base.Experimental ./experimental.jl:63
 [7] top-level scope
   @ experimental.jl:101

    nested task error: die
    Stacktrace:
     [1] error(s::String)
       @ Base ./error.jl:33
     [2] (::var"#20#23")()
       @ Main ./task.jl:406
```

**Comment by tro3** [comment]
Yup.  I had run across the disappearing exception problem while working on [Threadpools](https://github.com/tro3/ThreadPools.jl) and got interested in the lower-level stuff.  (I'm coming from an hardware description language background where concurrency is the norm.)  I had run a test on `Experimental.@sync` and convinced myself it wasn't handling the case, but after your feedback above and that from @StefanKarpinski and @JeffBezanson on #38916, I went back and found a bug and that it was.  In working on that PR, I meanderingly arrived at a structure very similar to that in Experimental, but with the Experimental implementation being superior.  IMO, you guys can move that Experimental version into the mainstream, and if you do, I've got a lockup test suite I'll be happy to PR.  Thanks for the help.
